<!DOCTYPE html>












  


<html class="theme-next mist use-motion" lang="zh-CN">
<head><meta name="generator" content="Hexo 3.8.0">
  <meta charset="UTF-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">












<meta http-equiv="Cache-Control" content="no-transform">
<meta http-equiv="Cache-Control" content="no-siteapp">






















<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css">

<link href="/css/main.css?v=6.4.0" rel="stylesheet" type="text/css">


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=6.4.0">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=6.4.0">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=6.4.0">


  <link rel="mask-icon" href="/images/logo.svg?v=6.4.0" color="#222">









<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Mist',
    version: '6.4.0',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: false,
    fastclick: false,
    lazyload: false,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>


  




  <meta name="description" content="一个实现基因组预测的实例如上一节所述，许多方法已被用于基因组预测。这里，将使用R包rrBLUP展示实现基因组预测的示例 (Endelman, 2011)。在演示中，将使用另一个R包BLR中的599个小麦品系的数据集，这些小麦系在1279个DArT标记上进行基因分型。如果你是语言新手，有许多在线资源可用于获取R的基本介绍 (Torfs &amp;amp; Brauer, 2014; Venables, Sm">
<meta property="og:type" content="article">
<meta property="og:title" content="水产育种中的基因组选择（翻译）(二)">
<meta property="og:url" content="flystar233.github.io/2018/11/13/Genomic-Selection-in-Aquaculture-Breeding-Programs2/index.html">
<meta property="og:site_name" content="rainsong">
<meta property="og:description" content="一个实现基因组预测的实例如上一节所述，许多方法已被用于基因组预测。这里，将使用R包rrBLUP展示实现基因组预测的示例 (Endelman, 2011)。在演示中，将使用另一个R包BLR中的599个小麦品系的数据集，这些小麦系在1279个DArT标记上进行基因分型。如果你是语言新手，有许多在线资源可用于获取R的基本介绍 (Torfs &amp;amp; Brauer, 2014; Venables, Sm">
<meta property="og:locale" content="zh-CN">
<meta property="og:image" content="/2018/11/13/Genomic-Selection-in-Aquaculture-Breeding-Programs2/number.png">
<meta property="og:updated_time" content="2019-04-17T09:06:08.292Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="水产育种中的基因组选择（翻译）(二)">
<meta name="twitter:description" content="一个实现基因组预测的实例如上一节所述，许多方法已被用于基因组预测。这里，将使用R包rrBLUP展示实现基因组预测的示例 (Endelman, 2011)。在演示中，将使用另一个R包BLR中的599个小麦品系的数据集，这些小麦系在1279个DArT标记上进行基因分型。如果你是语言新手，有许多在线资源可用于获取R的基本介绍 (Torfs &amp;amp; Brauer, 2014; Venables, Sm">
<meta name="twitter:image" content="/2018/11/13/Genomic-Selection-in-Aquaculture-Breeding-Programs2/number.png">



  <link rel="alternate" href="/rss.xml" title="rainsong" type="application/atom+xml">




  <link rel="canonical" href="flystar233.github.io/2018/11/13/Genomic-Selection-in-Aquaculture-Breeding-Programs2/">



<script type="text/javascript" id="page.configurations">
  CONFIG.page = {
    sidebar: "",
  };
</script>

  <title>水产育种中的基因组选择（翻译）(二) | rainsong</title>
  









  <noscript>
  <style type="text/css">
    .use-motion .motion-element,
    .use-motion .brand,
    .use-motion .menu-item,
    .sidebar-inner,
    .use-motion .post-block,
    .use-motion .pagination,
    .use-motion .comments,
    .use-motion .post-header,
    .use-motion .post-body,
    .use-motion .collection-title { opacity: initial; }

    .use-motion .logo,
    .use-motion .site-title,
    .use-motion .site-subtitle {
      opacity: initial;
      top: initial;
    }

    .use-motion {
      .logo-line-before i { left: initial; }
      .logo-line-after i { right: initial; }
    }
  </style>
</noscript>

</head>

<body itemscope="" itemtype="http://schema.org/WebPage" lang="zh-CN">

  
  
    
  

  <div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>
    <a href="https://github.com/flystar233" class="github-corner" aria-label="View source on Github"><svg width="80" height="80" viewbox="0 0 250 250" style="fill:#151513; color:#fff; position: absolute; top: 0; border: 0; right: 0;" aria-hidden="true"><path d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"/><path d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2" fill="currentColor" style="transform-origin: 130px 106px;" class="octo-arm"/><path d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z" fill="currentColor" class="octo-body"/></svg></a><style>.github-corner:hover .octo-arm{animation:octocat-wave 560ms ease-in-out}@keyframes octocat-wave{0%,100%{transform:rotate(0)}20%,60%{transform:rotate(-25deg)}40%,80%{transform:rotate(10deg)}}@media (max-width:500px){.github-corner:hover .octo-arm{animation:none}.github-corner .octo-arm{animation:octocat-wave 560ms ease-in-out}}</style>
    <header id="header" class="header" itemscope="" itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/" class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">rainsong</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
    
      
        <p class="site-subtitle">Everyone is a dreamer</p>
      
    
  </div>

  <div class="site-nav-toggle">
    <button aria-label="切换导航栏">
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>



<nav class="site-nav">
  
    <ul id="menu" class="menu">
      
        
        
        
          
          <li class="menu-item menu-item-home">
    <a href="/home/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-home"></i> <br>首页</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-about">
    <a href="/about/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-user"></i> <br>关于</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-categories">
    <a href="/categories/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-th"></i> <br>分类</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-archives">
    <a href="/archives/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-archive"></i> <br>归档</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-links">
    <a href="/links/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-calendar"></i> <br>链接</a>
  </li>

      
      
        <li class="menu-item menu-item-search">
          
            <a href="javascript:;" class="popup-trigger">
          
            
              <i class="menu-item-icon fa fa-search fa-fw"></i> <br>搜索</a>
        </li>
      
    </ul>
  

  
    

  

  
    <div class="site-search">
      
  <div class="popup search-popup local-search-popup">
  <div class="local-search-header clearfix">
    <span class="search-icon">
      <i class="fa fa-search"></i>
    </span>
    <span class="popup-btn-close">
      <i class="fa fa-times-circle"></i>
    </span>
    <div class="local-search-input-wrapper">
      <input autocomplete="off" placeholder="搜索..." spellcheck="false" type="text" id="local-search-input">
    </div>
  </div>
  <div id="local-search-result"></div>
</div>


    </div>
  
</nav>



  



</div>
    </header>

    


    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  

  <article class="post post-type-normal" itemscope="" itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="flystar233.github.io/2018/11/13/Genomic-Selection-in-Aquaculture-Breeding-Programs2/">

    <span hidden itemprop="author" itemscope="" itemtype="http://schema.org/Person">
      <meta itemprop="name" content="轩暮">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/add_fill.png">
    </span>

    <span hidden itemprop="publisher" itemscope="" itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="rainsong">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">水产育种中的基因组选择（翻译）(二)
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              

              
                
              

              <time title="创建时间：2018-11-13 10:54:46" itemprop="dateCreated datePublished" datetime="2018-11-13T10:54:46+08:00">2018-11-13</time>
            

            
              

              
                
                <span class="post-meta-divider">|</span>
                

                <span class="post-meta-item-icon">
                  <i class="fa fa-calendar-check-o"></i>
                </span>
                
                  <span class="post-meta-item-text">更新于</span>
                
                <time title="修改时间：2019-04-17 17:06:08" itemprop="dateModified" datetime="2019-04-17T17:06:08+08:00">2019-04-17</time>
              
            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope="" itemtype="http://schema.org/Thing"><a href="/categories/bioinfo/" itemprop="url" rel="index"><span itemprop="name">bioinfo</span></a></span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <h2 id="一个实现基因组预测的实例"><a href="#一个实现基因组预测的实例" class="headerlink" title="一个实现基因组预测的实例"></a>一个实现基因组预测的实例</h2><p>如上一节所述，许多方法已被用于基因组预测。这里，将使用R包<code>rrBLUP</code>展示实现基因组预测的示例 (Endelman, 2011)。在演示中，将使用另一个R包<code>BLR</code>中的599个小麦品系的数据集，这些小麦系在1279个DArT标记上进行基因分型。如果你是语言新手，有许多在线资源可用于获取R的基本介绍 (Torfs &amp; Brauer, 2014; Venables, Smith &amp; R Development Core Team, 2016)。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br></pre></td><td class="code"><pre><span class="line">install.packages(<span class="string">'rrBLUP'</span>)</span><br><span class="line">install.packages(<span class="string">'BLR'</span>)</span><br><span class="line"><span class="keyword">library</span>(BLR)</span><br><span class="line"><span class="keyword">library</span>(rrBLUP)</span><br><span class="line"></span><br><span class="line">rm(list=ls())</span><br><span class="line">set.seed(<span class="number">99</span>)  <span class="comment">#设置随机数</span></span><br><span class="line">data(wheat)  <span class="comment">#载入小麦数据</span></span><br><span class="line"><span class="comment">############################################### marker-based</span></span><br><span class="line">M &lt;- <span class="number">2</span>∗X-<span class="number">1</span>  <span class="comment">#将标记转换为&#123;−1，1&#125;</span></span><br><span class="line">pheno &lt;-Y[,<span class="number">1</span>]  <span class="comment">#取第一列表型数据</span></span><br><span class="line">geno &lt;- M   <span class="comment">#取基因型数据</span></span><br><span class="line">whichTest&lt;-sample(<span class="number">1</span>:length(pheno),<span class="number">100</span>)   <span class="comment">#随机取出100个样本表型数据的编号</span></span><br><span class="line">phenoTrain &lt;- pheno[-whichTest]   <span class="comment">#取出训练集表型数据</span></span><br><span class="line">phenoTest &lt;- pheno[whichTest]     <span class="comment">#取出测试集表型数据</span></span><br><span class="line">GenoTrain &lt;- as.matrix(geno[-whichTest, ])   <span class="comment">#取出训练集基因型数据</span></span><br><span class="line">genoTest &lt;- as.matrix(geno[whichTest, ])     <span class="comment">##取出测试集基因型数据</span></span><br><span class="line">markerEffects &lt;- mixed.solve(y=phenoTrain, Z=GenoTrain)$u   <span class="comment">#计算位点效应</span></span><br><span class="line">predictedGBV &lt;- genoTest%∗%markerEffects    <span class="comment">#计算育种值</span></span><br><span class="line">(predictionAcc &lt;- cor(predictedGBV, phenoTest))   <span class="comment">#计算育种精确度</span></span><br><span class="line"><span class="comment"># 0.5370625</span></span><br><span class="line"></span><br><span class="line"><span class="comment">############################################### kinship-based</span></span><br><span class="line">A1 &lt;-A.mat(M)   <span class="comment">#计算基因组关联矩阵</span></span><br><span class="line">rownames(A1) &lt;- <span class="number">1</span>:length(pheno)   <span class="comment">#替换A1矩阵行名</span></span><br><span class="line">yNA=pheno   <span class="comment">#取出表型</span></span><br><span class="line">yNA[whichTest] &lt;- <span class="literal">NA</span>   <span class="comment">#将测试集表型删除</span></span><br><span class="line">data1 &lt;-data.frame(y=yNA,gid=<span class="number">1</span>:length(pheno))   <span class="comment">#构建列表</span></span><br><span class="line">ans1 &lt;- kin.blup(data1,K=A1,geno=<span class="string">"gid"</span>,pheno=<span class="string">"y"</span>)   <span class="comment">#计算育种结果</span></span><br><span class="line">(cor(ans1$g[whichTest],Y[whichTest,<span class="number">1</span>]))   <span class="comment">#计算育种精确度</span></span><br><span class="line"><span class="comment"># 0.5370625</span></span><br></pre></td></tr></table></figure>
<h2 id="GS的几个重要考虑"><a href="#GS的几个重要考虑" class="headerlink" title="GS的几个重要考虑"></a>GS的几个重要考虑</h2><h3 id="需要多少动物进行基因分型？"><a href="#需要多少动物进行基因分型？" class="headerlink" title="需要多少动物进行基因分型？"></a>需要多少动物进行基因分型？</h3><p>GS 的准确性对与育种成功至关重要，它本身就是动植物育种的一个重要研究领域。除了选择统计模型外，GS 的准确性还取决于许多相互关联的因素，即基因组大小，标记密度，标记与 QTL之间的 LD，QTL 效应的数量和大小，有效群体大小，训练集的大小，训练与测试集动物的关系和性状遗传力。为了达到所需的精确度，训练集的动物数量主要取决于性状的遗传力估计和种群的有效规模。 (Ne) (Figure 2)。较低的遗传力估计将需要一个更大的训练集，同时，较大的Ne也需要相应的更大的训练样本。通常，更多的训练样本将为候选样本提供更准确的预测。例如， 荷斯坦奶牛需要有基因型和表型记录的3000-10000头公牛参考种群才能有0.7的基因组选择精确度。</p>
<img src="/2018/11/13/Genomic-Selection-in-Aquaculture-Breeding-Programs2/number.png">
<p>​        <strong>Figure 2</strong> 参考种群中所需的动物数量，以便为估计的GBV精确性达到0.7(Goddard &amp; Hayes, 2009)</p>
<h3 id="多少SNP位点才足够？"><a href="#多少SNP位点才足够？" class="headerlink" title="多少SNP位点才足够？"></a>多少SNP位点才足够？</h3><p>snp位点应该覆盖整个基因组，并且要有足够的密度以确保大多数QTL都在LD中，需要至少有一个标记在LD区间，因此，对标记效应的估计可以捕捉到最大的遗传变异。所需SNP的总数主要取决于基因组大小和基因组的LD范围。通常，相邻SNP标记之间的LD的 <code>r2 &gt;0.2</code> ，那SNP将会足够(Calus et al., 2008)。</p>
<p>在实践中，50 K SNP 为奶牛品种提供了较好的预测精度。将密度从50 K进一步增加到800 K，只会略微增加预测精度 (Khatkar et al., 2012)。然而，如果训练集和测试集动物是相关的，基因组关联矩阵（GRM）可用于使用 gBLUP 预测 GBV。GRM 可以用一个很小的 SNP 集精确地计算出来。</p>
<p>用中等密度 SNP 芯片进行基因分型的成本(例如，牛的50 K SNP芯片)大约是每一个 DNA 样本40到100美元。基因分型和测序的成本正在迅速下降，使这项技术应用于更多物种，成本效益更高。此外，基因型估算可以有效地提高基因型数据的密度，降低成本。基因型估算包括用高密度 SNP 集对一小部分群体进行基因分型，并用低成本、低密度 SNP 位点预测其他群体的高密度基因型，这些经过计算模拟的基因型虽然不是太准确，但是还可以用于 GS (Khatkar et al., 2012)。</p>
<h3 id="对基因组预测来说，最重要的因素是样本数还是SNPs数？"><a href="#对基因组预测来说，最重要的因素是样本数还是SNPs数？" class="headerlink" title="对基因组预测来说，最重要的因素是样本数还是SNPs数？"></a>对基因组预测来说，最重要的因素是样本数还是SNPs数？</h3><p>更多的记录和更紧密的标记间距提高了准确性。但是，当 SNP 的密度很小时（小于1000个SNPs），在训练集中增加更多的动物对于提高基因组预测的准确性变得更加重要。</p>
<h3 id="是否有可能预测不同品种-群体？"><a href="#是否有可能预测不同品种-群体？" class="headerlink" title="是否有可能预测不同品种/群体？"></a>是否有可能预测不同品种/群体？</h3><p>是否有可能预测不同品种/群体—也就是说，使用一个品种的训练动物为另一个品种做预测。GS已成功地应用于群体内育种值的预测。然而，GS 在杂交预测中的成功是有限的。例如，当计算不属于训练组的品种的基因组预测时，精确度接近于零，或非常低 (Kachman et al., 2013)。在这种情况下，使用几个具有中/大效应的 SNP 并采用贝叶斯方法可能会有帮助，因为这可能只包括那些跨品系/品种分离的 QTL。</p>
<h3 id="我们需要有关基因和基因功能的知识吗？"><a href="#我们需要有关基因和基因功能的知识吗？" class="headerlink" title="我们需要有关基因和基因功能的知识吗？"></a>我们需要有关基因和基因功能的知识吗？</h3><p>在实践中，大多数GS方法主要使用(非线性)线性模型估计的 SNP 效应，或者仅仅使用 GRM，因此它们没有利用基因、基因功能的特定知识，甚至没有基因组中的SNPs的精确定位。</p>
<h3 id="准确性会随着世代下降吗？"><a href="#准确性会随着世代下降吗？" class="headerlink" title="准确性会随着世代下降吗？"></a>准确性会随着世代下降吗？</h3><p>一般，基因组选择中GBVs的预测是利用标记与数量性状位点（QTL）之间的 LD 来实现的。然而，在实践中，尤其是在种内预测，基因组预测的可靠性取决于候选/测试中动物与训练/参考动物之间遗传关系的强弱。因此，当应用于下一代时，来自固定训练集的预测方程的精度会迅速下降。这通常意味着预测方程需要通过添加来自更近几代的动物来更新。</p>
<h3 id="使用GS会增加近亲繁殖吗？"><a href="#使用GS会增加近亲繁殖吗？" class="headerlink" title="使用GS会增加近亲繁殖吗？"></a>使用GS会增加近亲繁殖吗？</h3><p>较短的世代间隔和较高的选择强度将使近交系的年率提高。特别是在选择的标记附近近亲繁殖。这将需要在管理种群近亲繁殖方面给予更多的关注。利用基因组关系信息进行仔细的交配选择，如最小同祖先交配和最优贡献选择，可以控制近交系。</p>
<h2 id="水产养殖中的基因组选择"><a href="#水产养殖中的基因组选择" class="headerlink" title="水产养殖中的基因组选择"></a>水产养殖中的基因组选择</h2><p>水产养殖物种繁多，由于生命周期、繁殖力、有效种群规模和育种目标的不同，GS 在水产养殖中的潜力将因品种不同而有所不同。目前，以水产养殖品种最多的传统养殖项目主要采用群体选择和家系选择两种方式。同胞系测试是针对那些不能直接在候选样本上测量的性状进行的(例如，在生命后期或屠宰时记录到的性状、抗病性)。以家系为基础的选择只使用了一小部分(一半)的遗传变异，并导致近亲育种的增加。GS 还可以预测家族内部的遗传差异，因此可以利用所有的遗传变异。由于只使用很少的动物/家系而导致近亲繁殖的增加是水产养殖中的一个主要问题。此外，在传统的选择方案中，个体的标记、谱系的记录和单独家系的饲养也是困难和昂贵的。遗传标记的信息可以帮助应对其中的一些挑战。随着测序和基因分型成本的降低，遗传标记和 GS 在水产养殖中的应用越来越受到人们的重视。然而，SNP 集的开发和基因组资源的应用直到最近才在几个水产养殖物种中进行，以此，在实际 GS 实施结果的信息是不全面的。已发表的报告主要是基于 GS 使用模拟数据的效率。所有的模拟研究表明，使用 GS 来进行水产养殖育种可以提高在生产(连续型)和疾病(二分型)位点的选择和遗传增益的准确性。(Sonesson &amp; Meuwissen, 2009)；(Nielsen, Sonesson &amp; Meuwissen, 2011)； (Lillehammer,Meuwissen &amp; Sonesson, 2013)，与传统的同胞系检测相比，近亲繁殖减少了81%(Sonesson &amp; Meuwissen, 2009)。通过结合传统的 BLUP 家系育种值和基于低密度基因分型而不影响遗传增益的家系内育种值，也可以降低基因分型的成本(Lillehammer et al., 2013)。</p>
<p>尽管基因分型和测序成本下降，由于大量样本的养殖和基因分型，基因组选择策略仍然需要很多的成本。然而，部分额外成本可以从增加的遗传收益中弥补而且与传统的选择方案相比，对同胞系样本测试的依赖更少。大多数水产养殖物种中个体动物的价值一般较低（与牛相比）。然而，由于大多数水产养殖物种的繁殖力很高，遗传增益的影响因遗传收益从繁殖核心立即转移到商业池塘而扩大。然而，利用GS优化育种方案还需要进一步的研究，比如在训练和选择中确定最佳的动物数量，标记密度，最小化近亲繁殖最大化长期遗传收益，与传统方案的比较的经济分析。GS 除了用于性状外，遗传标记在水产养殖中还可以提供比传统方法更多的优势，例如，性别相关标记的使用可以生产单性别的后代(Robinson et al., 2014)；基因渗入可以用来在种群中引入理想的基因，例如，通过反复回交和标记辅助选择，将对特定疾病的抗性从本地菌株转移到商品菌株(Odegard et al., 2009)；标记可用于监测和繁殖来自群体的任何有害突变；遗传物质的可操纵性来保护育种者。</p>
<p>总之，随着基因组技术的快速发展和基因组信息获取成本的不断降低，全基因组选择在提高盈利能力和在个体数增加依旧保持遗传变异性的同时，对提高水产养殖品种的食品生产水平具有很大的潜力。</p>
<h2 id="致谢"><a href="#致谢" class="headerlink" title="致谢"></a>致谢</h2><p>十分感谢 Dr. Gerhard Moser 对此手稿的建议。</p>
<h2 id="参考"><a href="#参考" class="headerlink" title="参考"></a>参考</h2><p>Calus, M.P., Meuwissen, T.H., de Roos, A.P. and Veerkamp, R.F. (2008) Accuracy of genomic selection using different methods to define haplotypes. Genetics, 178,553–561.<br>de Los Campos, G., Hickey, J.M., Pong-Wong, R. et al. (2013) Whole-genome regression and prediction methods applied to plant and animal breeding. Genetics, 193, 327–345.<br>de Los Campos, G., Naya, H., Gianola, D. et al. (2009) Predicting quantitative traits with regression models for dense molecular markers and pedigree. Genetics, 182, 375–385.<br>Endelman, J.B. (2011) Ridge regression and other kernels for genomic selection with R package rrBLUP. Plant Genome, 4, 250–255.<br>Erbe, M., Hayes, B.J., Matukumalli, L.K. et al. (2012) Improving accuracy of genomic predictions within and between dairy cattle breeds with imputed high-density single<br>nucleotide polymorphism panels. Journal of Dairy Science, 95, 4114–4129.<br>Falconer, D.S. and Mackay, T.F.C. (1996) Introduction to quantitative genetics, 4th edn, Longman, Essex, UK.<br>Gianola, D. and van Kaam, J.B. (2008) Reproducing kernel Hilbert spaces regression methods for genomic assisted prediction of quantitative traits. Genetics, 178,2289–2303.<br>Goddard, M.E. and Hayes, B.J. (2009) Mapping genes for complex traits in domestic animals and their use in breeding programmes. Nature Reviews Genetics, 10, 381–391.<br>Gonzalez-Camacho, J.M., de Los Campos, G., Perez, P. et al. (2012) Genome-enabled prediction of genetic values using radial basis function neural networks. Theoretical and Applied Genetics, 125, 759–771.<br>Gonzalez-Recio, O., Weigel, K.A., Gianola, D. et al. (2010) L2-Boosting algorithm applied to high-dimensional problems in genomic selection. Genetics Research (Cambridge), 92,227–237.<br>Habier, D., Fernando, R.L. and Dekkers, J.C. (2007) The impact of genetic relationship information on genome-assisted breeding values. Genetics, 177, 2389–2397</p>
<p>Habier, D., Fernando, R.L., Kizilkaya, K. and Garrick, D.J. (2011) Extension of the Bayesian alphabet for genomic selection. BMC Bioinformatics, 12, 186.</p>
<p>Heslot, N., Yang, H.P., Sorrells, M.E. and Jannink, J.L. (2012) Genomic selection in plant breeding: a comparison of models. Crop Science, 52 (1), 146–160.</p>
<p>Jannink, J.L., Lorenz, A.J. and Iwata, H. (2010) Genomic selection in plant breeding: from theory to practice. Briefings in Functional Genomics, 9, 166–177.<br>Kachman, S.D., Spangler, M.L., Bennett, G.L. et al. (2013) Comparison of molecular breeding values based on within- and across-breed training in beef cattle. Genetics Selection Evolution, 45, 30.<br>Khatkar, M.S., Moser, G., Hayes, B.J. and Raadsma, H.W. (2012) Strategies and utility of imputed SNP genotypes for genomic analysis in dairy cattle. BMC Genomics, 13, 538.<br>Khatkar, M.S., Zenger, K.R., Hobbs, M. et al. (2007) A primary assembly of a bovine haplotype block map based on a 15,036-single-nucleotide polymorphism panel genotyped in Holstein-Friesian cattle. Genetics, 176, 763–772.<br>Lillehammer, M., Meuwissen, T.H. and Sonesson, A.K. (2013) A low-marker density implementation of genomic selection in aquaculture using within-family genomic breeding values. Genetics Selection Evolution, 45, 39.<br>Maenhout, S., De Baets, B. and Haesaert, G. (2010) Prediction of maize single-cross hybrid performance: support vector machine regression versus best linear prediction. Theoretical and Applied Genetics, 120, 415–427.<br>Meuwissen, T.H., Hayes, B.J. and Goddard, M.E. (2001) Prediction of total genetic value using genome-wide dense marker maps. Genetics, 157, 1819–1829.<br>Moser, G., Tier, B., Crump, R.E. et al. (2009) A comparison of five methods to predictgenomic breeding values of dairy bulls from genome-wide SNP markers. Genetics Selection Evolution, 41, 56.<br>Nejati-Javaremi, A., Smith, C. and Gibson, J.P. (1997) Effect of total allelic relationship on accuracy of evaluation and response to selection. Journal of Animal Science, 75,1738–1745.<br>Nielsen, H.M., Sonesson, A.K. and Meuwissen, T.H. (2011) Optimum contribution selection using traditional best linear unbiased prediction and genomic breeding values in aquaculture breeding schemes. Journal of Animal Science, 89, 630–638.<br>Odegard, J., Yazdi, M.H., Sonesson, A.K. and Meuwissen, T.H. (2009) Incorporating desirable genetic characteristics from an inferior into a superior population using genomic selection. Genetics, 181, 737–745.<br>Ogutu, J.O., Piepho, H.P. and Schulz-Streeck, T. (2011) A comparison of random forests, boosting and support vector machines for genomic selection. BMC Proceedings, 5 (3), S11.<br>Pryce, J.E., Goddard, M.E., Raadsma, H.W. and Hayes, B.J. (2010) Deterministic models of breeding scheme designs that incorporate genomic selection. Journal of Dairy Science,93, 5455–5466.<br>Robinson, N.A., Gopikrishna, G., Baranski, M. et al. (2014) QTL for white spot syndrome virus resistance and the sex-determining locus in the Indian black tiger shrimp (Penaeus monodon). BMC Genomics, 15, 731.<br>Schaeffer, L.R. (2006) Strategy for applying genome-wide selection in dairy cattle. Journal of Animal Breeding and Genetics, 123, 218–223.<br>Sonesson, A.K. and Meuwissen, T.H.E. (2009) Testing strategies for genomic selection in aquaculture breeding programs. Genetics Selection Evolution, 41, 1.</p>
<p>Torfs, P. and Brauer, C. (2014). “A (very) short introduction to R,” Hydrology and Quantitative Water Management Group, Wageningen University, The Netherlands,available at <a href="https://cran.r-project.org/doc/contrib/Torfs+Brauer-Short-R-Intro.pdf" target="_blank" rel="noopener">https://cran.r-project.org/doc/contrib/Torfs+Brauer-Short-R-Intro.pdf</a></p>
<p>Tsai, H.Y., Hamilton, A., Tinch, A.E. et al. (2015) Genome wide association and genomic prediction for growth traits in juvenile farmed Atlantic salmon using a high density SNP array. BMC Genomics, 16, 969.<br>VanRaden, P.M. (2008) Efficient methods to compute genomic predictions. Journal of Dairy Science, 91，4414–4423.<br>VanRaden, P.M., Van Tassell, C.P., Wiggans, G.R. et al. (2009) Invited review: reliability of genomic predictions for North American Holstein bulls. Journal of Dairy Science, 92,16–24.<br>Vazquez, A.I., de los Campos, G., Klimentidis, Y.C. et al. (2012) A comprehensive genetic approach for improving prediction of skin cancer risk in humans. Genetics, 192,1493–1502.<br>Venables, W.N., Smith, D.M. and the R Core Team. (2016). An introduction to R.<br>Notes on R: A programming environment for data analysis and graphics. Version 3.2.4 (2016-03-10) Available: <a href="https://cran.r-project.org/doc/manuals/R-intro.pdf" target="_blank" rel="noopener">https://cran.r-project.org/doc/manuals/R-intro.pdf</a>. Accessed 15 April 2016.</p>

      
    </div>

    

    
    
    

    

    

    

    <footer class="post-footer">
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2018/11/08/Genomic-Selection-in-Aquaculture-Breeding-Programs/" rel="next" title="水产育种中的基因组选择（翻译）(一)">
                <i class="fa fa-chevron-left"></i> 水产育种中的基因组选择（翻译）(一)
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2018/11/20/sam-header/" rel="prev" title="sam文件header全面解析">
                sam文件header全面解析 <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>


  </div>


          </div>
          

  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            文章目录
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            站点概览
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope="" itemtype="http://schema.org/Person">
            
              <img class="site-author-image" itemprop="image" src="/images/add_fill.png" alt="轩暮">
            
              <p class="site-author-name" itemprop="name">轩暮</p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          
            <nav class="site-state motion-element">
              
                <div class="site-state-item site-state-posts">
                
                  <a href="/archives/">
                
                    <span class="site-state-item-count">36</span>
                    <span class="site-state-item-name">日志</span>
                  </a>
                </div>
              

              
                
                
                <div class="site-state-item site-state-categories">
                  <a href="/categories/index.html">
                    
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">7</span>
                    <span class="site-state-item-name">分类</span>
                  </a>
                </div>
              

              
            </nav>
          

          
            <div class="feed-link motion-element">
              <a href="/rss.xml" rel="alternate">
                <i class="fa fa-rss"></i>
                RSS
              </a>
            </div>
          

          
            <div class="links-of-author motion-element">
              
                <span class="links-of-author-item">
                  <a href="https://github.com/flystar233" target="_blank" title="GitHub"><i class="fa fa-fw fa-github"></i>GitHub</a>
                  
                </span>
              
            </div>
          

          
          

          
          

          
            
          
          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#一个实现基因组预测的实例"><span class="nav-number">1.</span> <span class="nav-text">一个实现基因组预测的实例</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#GS的几个重要考虑"><span class="nav-number">2.</span> <span class="nav-text">GS的几个重要考虑</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#需要多少动物进行基因分型？"><span class="nav-number">2.1.</span> <span class="nav-text">需要多少动物进行基因分型？</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#多少SNP位点才足够？"><span class="nav-number">2.2.</span> <span class="nav-text">多少SNP位点才足够？</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#对基因组预测来说，最重要的因素是样本数还是SNPs数？"><span class="nav-number">2.3.</span> <span class="nav-text">对基因组预测来说，最重要的因素是样本数还是SNPs数？</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#是否有可能预测不同品种-群体？"><span class="nav-number">2.4.</span> <span class="nav-text">是否有可能预测不同品种/群体？</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#我们需要有关基因和基因功能的知识吗？"><span class="nav-number">2.5.</span> <span class="nav-text">我们需要有关基因和基因功能的知识吗？</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#准确性会随着世代下降吗？"><span class="nav-number">2.6.</span> <span class="nav-text">准确性会随着世代下降吗？</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#使用GS会增加近亲繁殖吗？"><span class="nav-number">2.7.</span> <span class="nav-text">使用GS会增加近亲繁殖吗？</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#水产养殖中的基因组选择"><span class="nav-number">3.</span> <span class="nav-text">水产养殖中的基因组选择</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#致谢"><span class="nav-number">4.</span> <span class="nav-text">致谢</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#参考"><span class="nav-number">5.</span> <span class="nav-text">参考</span></a></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright">&copy; <span itemprop="copyrightYear">2019</span>
  <span class="with-love" id="animate">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">轩暮</span>

  

  
</div>

















        








        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    
	
    

    
  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>


























  
  
    <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>
  

  
  
    <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>
  

  
  
    <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>
  


  


  <script type="text/javascript" src="/js/src/utils.js?v=6.4.0"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=6.4.0"></script>



  
  

  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=6.4.0"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=6.4.0"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=6.4.0"></script>



  



  










  

  <script type="text/javascript">
    // Popup Window;
    var isfetched = false;
    var isXml = true;
    // Search DB path;
    var search_path = "search.xml";
    if (search_path.length === 0) {
      search_path = "search.xml";
    } else if (/json$/i.test(search_path)) {
      isXml = false;
    }
    var path = "/" + search_path;
    // monitor main search box;

    var onPopupClose = function (e) {
      $('.popup').hide();
      $('#local-search-input').val('');
      $('.search-result-list').remove();
      $('#no-result').remove();
      $(".local-search-pop-overlay").remove();
      $('body').css('overflow', '');
    }

    function proceedsearch() {
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay"></div>')
        .css('overflow', 'hidden');
      $('.search-popup-overlay').click(onPopupClose);
      $('.popup').toggle();
      var $localSearchInput = $('#local-search-input');
      $localSearchInput.attr("autocapitalize", "none");
      $localSearchInput.attr("autocorrect", "off");
      $localSearchInput.focus();
    }

    // search function;
    var searchFunc = function(path, search_id, content_id) {
      'use strict';

      // start loading animation
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay">' +
          '<div id="search-loading-icon">' +
          '<i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>' +
          '</div>' +
          '</div>')
        .css('overflow', 'hidden');
      $("#search-loading-icon").css('margin', '20% auto 0 auto').css('text-align', 'center');

      

      $.ajax({
        url: path,
        dataType: isXml ? "xml" : "json",
        async: true,
        success: function(res) {
          // get the contents from search data
          isfetched = true;
          $('.popup').detach().appendTo('.header-inner');
          var datas = isXml ? $("entry", res).map(function() {
            return {
              title: $("title", this).text(),
              content: $("content",this).text(),
              url: $("url" , this).text()
            };
          }).get() : res;
          var input = document.getElementById(search_id);
          var resultContent = document.getElementById(content_id);
          var inputEventFunction = function() {
            var searchText = input.value.trim().toLowerCase();
            var keywords = searchText.split(/[\s\-]+/);
            if (keywords.length > 1) {
              keywords.push(searchText);
            }
            var resultItems = [];
            if (searchText.length > 0) {
              // perform local searching
              datas.forEach(function(data) {
                var isMatch = false;
                var hitCount = 0;
                var searchTextCount = 0;
                var title = data.title.trim();
                var titleInLowerCase = title.toLowerCase();
                var content = data.content.trim().replace(/<[^>]+>/g,"");
                
                var contentInLowerCase = content.toLowerCase();
                var articleUrl = decodeURIComponent(data.url);
                var indexOfTitle = [];
                var indexOfContent = [];
                // only match articles with not empty titles
                if(title != '') {
                  keywords.forEach(function(keyword) {
                    function getIndexByWord(word, text, caseSensitive) {
                      var wordLen = word.length;
                      if (wordLen === 0) {
                        return [];
                      }
                      var startPosition = 0, position = [], index = [];
                      if (!caseSensitive) {
                        text = text.toLowerCase();
                        word = word.toLowerCase();
                      }
                      while ((position = text.indexOf(word, startPosition)) > -1) {
                        index.push({position: position, word: word});
                        startPosition = position + wordLen;
                      }
                      return index;
                    }

                    indexOfTitle = indexOfTitle.concat(getIndexByWord(keyword, titleInLowerCase, false));
                    indexOfContent = indexOfContent.concat(getIndexByWord(keyword, contentInLowerCase, false));
                  });
                  if (indexOfTitle.length > 0 || indexOfContent.length > 0) {
                    isMatch = true;
                    hitCount = indexOfTitle.length + indexOfContent.length;
                  }
                }

                // show search results

                if (isMatch) {
                  // sort index by position of keyword

                  [indexOfTitle, indexOfContent].forEach(function (index) {
                    index.sort(function (itemLeft, itemRight) {
                      if (itemRight.position !== itemLeft.position) {
                        return itemRight.position - itemLeft.position;
                      } else {
                        return itemLeft.word.length - itemRight.word.length;
                      }
                    });
                  });

                  // merge hits into slices

                  function mergeIntoSlice(text, start, end, index) {
                    var item = index[index.length - 1];
                    var position = item.position;
                    var word = item.word;
                    var hits = [];
                    var searchTextCountInSlice = 0;
                    while (position + word.length <= end && index.length != 0) {
                      if (word === searchText) {
                        searchTextCountInSlice++;
                      }
                      hits.push({position: position, length: word.length});
                      var wordEnd = position + word.length;

                      // move to next position of hit

                      index.pop();
                      while (index.length != 0) {
                        item = index[index.length - 1];
                        position = item.position;
                        word = item.word;
                        if (wordEnd > position) {
                          index.pop();
                        } else {
                          break;
                        }
                      }
                    }
                    searchTextCount += searchTextCountInSlice;
                    return {
                      hits: hits,
                      start: start,
                      end: end,
                      searchTextCount: searchTextCountInSlice
                    };
                  }

                  var slicesOfTitle = [];
                  if (indexOfTitle.length != 0) {
                    slicesOfTitle.push(mergeIntoSlice(title, 0, title.length, indexOfTitle));
                  }

                  var slicesOfContent = [];
                  while (indexOfContent.length != 0) {
                    var item = indexOfContent[indexOfContent.length - 1];
                    var position = item.position;
                    var word = item.word;
                    // cut out 100 characters
                    var start = position - 20;
                    var end = position + 80;
                    if(start < 0){
                      start = 0;
                    }
                    if (end < position + word.length) {
                      end = position + word.length;
                    }
                    if(end > content.length){
                      end = content.length;
                    }
                    slicesOfContent.push(mergeIntoSlice(content, start, end, indexOfContent));
                  }

                  // sort slices in content by search text's count and hits' count

                  slicesOfContent.sort(function (sliceLeft, sliceRight) {
                    if (sliceLeft.searchTextCount !== sliceRight.searchTextCount) {
                      return sliceRight.searchTextCount - sliceLeft.searchTextCount;
                    } else if (sliceLeft.hits.length !== sliceRight.hits.length) {
                      return sliceRight.hits.length - sliceLeft.hits.length;
                    } else {
                      return sliceLeft.start - sliceRight.start;
                    }
                  });

                  // select top N slices in content

                  var upperBound = parseInt('1');
                  if (upperBound >= 0) {
                    slicesOfContent = slicesOfContent.slice(0, upperBound);
                  }

                  // highlight title and content

                  function highlightKeyword(text, slice) {
                    var result = '';
                    var prevEnd = slice.start;
                    slice.hits.forEach(function (hit) {
                      result += text.substring(prevEnd, hit.position);
                      var end = hit.position + hit.length;
                      result += '<b class="search-keyword">' + text.substring(hit.position, end) + '</b>';
                      prevEnd = end;
                    });
                    result += text.substring(prevEnd, slice.end);
                    return result;
                  }

                  var resultItem = '';

                  if (slicesOfTitle.length != 0) {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + highlightKeyword(title, slicesOfTitle[0]) + "</a>";
                  } else {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + title + "</a>";
                  }

                  slicesOfContent.forEach(function (slice) {
                    resultItem += "<a href='" + articleUrl + "'>" +
                      "<p class=\"search-result\">" + highlightKeyword(content, slice) +
                      "...</p>" + "</a>";
                  });

                  resultItem += "</li>";
                  resultItems.push({
                    item: resultItem,
                    searchTextCount: searchTextCount,
                    hitCount: hitCount,
                    id: resultItems.length
                  });
                }
              })
            };
            if (keywords.length === 1 && keywords[0] === "") {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-search fa-5x" /></div>'
            } else if (resultItems.length === 0) {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-frown-o fa-5x" /></div>'
            } else {
              resultItems.sort(function (resultLeft, resultRight) {
                if (resultLeft.searchTextCount !== resultRight.searchTextCount) {
                  return resultRight.searchTextCount - resultLeft.searchTextCount;
                } else if (resultLeft.hitCount !== resultRight.hitCount) {
                  return resultRight.hitCount - resultLeft.hitCount;
                } else {
                  return resultRight.id - resultLeft.id;
                }
              });
              var searchResultList = '<ul class=\"search-result-list\">';
              resultItems.forEach(function (result) {
                searchResultList += result.item;
              })
              searchResultList += "</ul>";
              resultContent.innerHTML = searchResultList;
            }
          }

          if ('auto' === 'auto') {
            input.addEventListener('input', inputEventFunction);
          } else {
            $('.search-icon').click(inputEventFunction);
            input.addEventListener('keypress', function (event) {
              if (event.keyCode === 13) {
                inputEventFunction();
              }
            });
          }

          // remove loading animation
          $(".local-search-pop-overlay").remove();
          $('body').css('overflow', '');

          proceedsearch();
        }
      });
    }

    // handle and trigger popup window;
    $('.popup-trigger').click(function(e) {
      e.stopPropagation();
      if (isfetched === false) {
        searchFunc(path, 'local-search-input', 'local-search-result');
      } else {
        proceedsearch();
      };
    });

    $('.popup-btn-close').click(onPopupClose);
    $('.popup').click(function(e){
      e.stopPropagation();
    });
    $(document).on('keyup', function (event) {
      var shouldDismissSearchPopup = event.which === 27 &&
        $('.search-popup').is(':visible');
      if (shouldDismissSearchPopup) {
        onPopupClose();
      }
    });
  </script>





  

  

  

  

  
  

  

  

  

  

  

</body>
</html>
